File: C_resources/sudoku.c
#include <stdio.h>
#define boolean int
#define TRUE 1
#define FALSE 0

boolean valid(int vec[9])
{
    int i;
    boolean flag[9] = {0};

    for(i=0; i<9; i++) {
        if (vec[i] != 0) {
            if (flag[vec[i]-1]) 
                return FALSE;
            else
                flag[vec[i]-1] = TRUE;
        }
    }
    return TRUE;
}

boolean check_sudoku(int sudoku[9][9], int row, int col)
{
    int vec[9];
    int row_sta, row_end;
    int col_sta, col_end;
    int i, j, k=0;

    row_sta = row/3*3;
    row_end = row_sta + 3;
    col_sta = col/3*3;
    col_end = col_sta + 3;

    // check row
    if (!valid(sudoku[row]))
        return FALSE;

    // check colomn
    for (i = 0; i < 9; i++) {
        vec[i] = sudoku[i][col];
    }
    if (!valid(vec))
        return FALSE;

    for(i=row_sta; i<row_end; i++) {
        for(j=col_sta; j<col_end; j++) {
            vec[k++] = sudoku[i][j];
        }
    }
    if (!valid(vec))
        return FALSE;
    
    return TRUE;
}

boolean find_first_empty(int sudoku[9][9], int* row, int* col)
{
    int i, j;
    for(i=0; i<9; i++) {
        for(j=0; j<9; j++) {
            if (sudoku[i][j] == 0) {
                *row = i;
                *col = j;
                return TRUE;
            }
        }
    }
    return FALSE;
}

void show_sudoku(int sudoku[9][9])
{
    int i, j;

    printf("\nSolution:\n");
    for(i=0; i<9; i++) {
        for(j=0; j<9; j++) {
            printf("%d ", sudoku[i][j]);
        }
        printf("\n");
    }
}

boolean solve_sudoku(int sudoku[9][9], int level)
{
    int i, j, k;
    boolean result = FALSE;
    
    if (!find_first_empty(sudoku, &i, &j)) {
        return TRUE;
    }
        
    for (k = 1; k <= 9; k++) {
        sudoku[i][j] = k;
        printf("\n%d: (%d,%d) -> %d",level, i, j, k);
        if (check_sudoku(sudoku, i, j) && solve_sudoku(sudoku, level+1)) {
            return TRUE;
        }
        else {
            sudoku[i][j] = 0;
        }
    }

    return FALSE;
}


int main()
{
    int  sudoku[9][9] = {{9,0,0,0,0,0,3,0,7},
                         {5,0,6,0,9,0,0,1,0},
                         {0,0,3,0,6,0,0,0,5},
                         {0,0,1,0,0,9,0,3,6},
                         {4,0,0,0,0,5,0,9,1},
                         {0,9,0,8,0,0,0,0,0},
                         {3,6,0,0,0,4,0,0,0},
                         {0,4,0,0,5,0,6,0,0},
                         {0,0,0,0,0,0,5,8,0}};

    if (solve_sudoku(sudoku, 0)) {
        show_sudoku(sudoku);
    }
    else {
        printf("\nFail");
    }

    return 1;
}

File: Python_resources/sudoku.py
import string

N = 9

def valid(vec):
    flag = [False for i in range(N)]
    for i in range(N):
        curr_char = vec[i]
        if (curr_char == '.'):
            continue
        curr_val = string.atoi(curr_char)
        if (flag[curr_val-1]):
            return False
        else:
            flag[curr_val-1] = True
    return True
            
def check_valid(board, x, y):
    vec = [0] * N
    #check row
    if (not valid(board[x])):
        return False

    #check col
    for i in range(N):
        vec[i] = board[i][y]
    if (not valid(vec)):
        return False

    #check block
    row_sta = x/3 * 3
    col_sta = y/3 * 3
    n = 0
    for i in range(row_sta, row_sta+3):
        for j in range(col_sta, col_sta+3):
            vec[n] = board[i][j]
            n = n + 1
    if (not valid(vec)):
        return False

    return True

    
def find_first_empty(board):
    for i in range(N):
        for j in range(N):
            if (board[i][j] == '.'):
                return (i, j)
    else:
        return (-1, -1)

def sudoku_print(board):
    for i in range(N):
        for j in range(N):
            print "%d  " % string.atoi(board[i][j]),
        print "\n"

def sudoku_solver(board, level=0):
    (x, y) = find_first_empty(board)
    if (x == -1 and y == -1):
        print "Solution found !\n"
        sudoku_print(board)
        return True

    #print level, x, y
    for val in range(1, N+1):        
        board[x][y] = '%d' % val
        if (check_valid(board, x, y) and
            sudoku_solver(board, level+1)):
            return True
        else:
            board[x][y] = '.'

    return False
    

board = [['.' for col in range(9)] for row in range(9)]
board[0] = "5 3 . . 7 . . . .".split()
board[1] = "6 . . 1 9 5 . . .".split()
board[2] = ". 9 8 . . . . 6 .".split()
board[3] = "8 . . . 6 . . . 3".split()
board[4] = "4 . . 8 . 3 . . 1".split()
board[5] = "7 . . . 2 . . . 6".split()
board[6] = ". 6 . . . . 2 8 .".split()
board[7] = ". . . 4 1 9 . . 5".split()
board[8] = ". . . . 8 . . 7 9".split()

sudoku_solver(board)

